### ISSUE TITLE
Infinite construction loop

### ISSUE THREAD

**Issue by JackTemaki** [rwth-i6/returnn#223]
After upgrading to the newest master from commit 4fa19f27718c16e81f111dc62e887cd677ecb20f, i encounter the following problem when constructing the network:

```
  File "/u/rossenbach/src/crnn_synthesis_master/TFNetworkRecLayer.py", line 1242, in _construct_template
    line: assert loop_limit >= 0, (
            ("We keep iterating over the network template construction.\n"
             "We have these partially finished layers:\n%s\n"
             "And these finished layers:\n%s\n"
             "And these recent changes in the last loop iteration:\n%s") % (
              pformat(ConstructCtx.partially_finished),
              pformat([
                layer for _, layer in sorted(self.layer_data_templates.items())
                if layer not in ConstructCtx.partially_finished]),
              pformat(recent_changes)))
    locals:
      loop_limit = <local> -1
      pformat = <local> <function pformat at 0x7fd63c6e6158>
      ConstructCtx = <local> <class 'TFNetworkRecLayer._SubnetworkRecCell._construct_template.<locals>.ConstructCtx'>
      ConstructCtx.partially_finished = <local> [<_TemplateLayer(LinearLayer)(:template:linear) 'output' out_type=Data(shape=(240,), time_dim_axis=None, batch_shape_meta=[B,F|240]) (construction stack None)>, <_TemplateLayer(RnnCellLayer)(:template:rnn_cell) 'decoder_2' out_type=Data(shape=(640,), time_dim_axis=None, batch_shape_meta=[B,F|640]..., len = 15
      layer = <local> <_TemplateLayer(EvalLayer)(:template:eval) 'entropy' out_type=Data(shape=(1, None), time_dim_axis=2, feature_dim_axis=1, available_for_inference=False, batch_shape_meta=[B,F|1,T|'time:var:extern_data:classes']) (construction stack None)>
      _ = <not found>
      sorted = <builtin> <built-in function sorted>
      self = <local> <_SubnetworkRecCell of None>
      self.layer_data_templates = <local> {'att_energy': <_TemplateLayer(LinearLayer)(:template:linear) 'att_energy' out_type=Data(shape=(None, 1), batch_dim_axis=1, available_for_inference=False, batch_shape_meta=[T|'time:var:extern_data:classes',B,F|1]) (construction stack 'att_weights')>, 'entropy': <_TemplateLayer(EvalLayer)(:templat..., len = 20
      self.layer_data_templates.items = <local> <built-in method items of dict object at 0x7fd568568708>
      recent_changes = <local> [(Data(name='location_feedback_transformed_output', shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|'time:var:extern_data:classes',F|128]), Data(name='location_feedback_transformed_output', shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|?,F|128]))]
AssertionError: We keep iterating over the network template construction.
```

I found that the issue is related to an infinite loop over this update during construction (local variable `recent_changes`):
```
[(Data(name='location_feedback_transformed_output', shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|'time:var:extern_data:classes',F|128]),
  Data(name='location_feedback_transformed_output', shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|?,F|128]))]
```

What could be the reason for this error?

**Comment by albertz** [comment]
Can you pinpoint the commit since when this happens more exactly? But I assume this is due to the more correct template construction, which does some extra checks. So in most cases, there was already a problem before but it was not detected. Maybe also the problem does not have an effect in the end for your case (although this is not clear).

It will probably help when you already prepare a simple test case which reproduces this. Try to reduce it as much as possible.

It would help if you post the full assertion message.

What kind of layer is this? Probably the `get_out_data_from_opts` of that layer is slightly buggy.

The question is why is it `[B,T|'time:var:extern_data:classes',F|128]` and then later `[B,T|?,F|128]` again.

**Comment by JackTemaki** [comment]
Yes, as expected the error occurs with 4bc5f842cd8c0aaf16d39d7a12902051c3a475d0 / "RecLayer template construction, more correct partially finished logic". The Layer is only a linear layer, i will try do reduce the network.

The full assertion message is:
```
  File "/u/rossenbach/src/crnn_synthesis_master/TFNetworkRecLayer.py", line 1241, in _construct_template
    line: assert loop_limit >= 0, (
            ("We keep iterating over the network template construction.\n"
             "We have these partially finished layers:\n%s\n"
             "And these finished layers:\n%s\n"
             "And these recent changes in the last loop iteration:\n%s") % (
              pformat(ConstructCtx.partially_finished),
              pformat([
                layer for _, layer in sorted(self.layer_data_templates.items())
                if layer not in ConstructCtx.partially_finished]),
              pformat(recent_changes)))
    locals:
      loop_limit = <local> -1
      pformat = <local> <function pformat at 0x7f81641f5158>
      ConstructCtx = <local> <class 'TFNetworkRecLayer._SubnetworkRecCell._construct_template.<locals>.ConstructCtx'>
      ConstructCtx.partially_finished = <local> [<_TemplateLayer(LinearLayer)(:template:linear) 'output' out_type=Data(shape=(240,), time_dim_axis=None, batch_shape_meta=[B,F|240]) (construction stack None)>, <_TemplateLayer(RnnCellLayer)(:template:rnn_cell) 'decoder_2' out_type=Data(shape=(640,), time_dim_axis=None, batch_shape_meta=[B,F|640]..., len = 15
      layer = <local> <_TemplateLayer(EvalLayer)(:template:eval) 'entropy' out_type=Data(shape=(1, None), time_dim_axis=2, feature_dim_axis=1, available_for_inference=False, batch_shape_meta=[B,F|1,T|'time:var:extern_data:classes']) (construction stack None)>
      _ = <not found>
      sorted = <builtin> <built-in function sorted>
      self = <local> <_SubnetworkRecCell of None>
      self.layer_data_templates = <local> {'location_feedback_transformed': <_TemplateLayer(LinearLayer)(:template:linear) 'location_feedback_transformed' out_type=Data(shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|?,F|128]) (construction stack 'att_energy_in')>, 'pre_slice': <_TemplateLayer(SliceLayer)(:templat..., len = 20
      self.layer_data_templates.items = <local> <built-in method items of dict object at 0x7f80903c7bc8>
      recent_changes = <local> [(Data(name='location_feedback_transformed_output', shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|'time:var:extern_data:classes',F|128]), Data(name='location_feedback_transformed_output', shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|?,F|128]))]
AssertionError: We keep iterating over the network template construction.
We have these partially finished layers:
[<_TemplateLayer(LinearLayer)(:template:linear) 'output' out_type=Data(shape=(240,), time_dim_axis=None, batch_shape_meta=[B,F|240]) (construction stack None)>,
 <_TemplateLayer(RnnCellLayer)(:template:rnn_cell) 'decoder_2' out_type=Data(shape=(640,), time_dim_axis=None, batch_shape_meta=[B,F|640]) (construction stack 'output')>,
 <_TemplateLayer(RnnCellLayer)(:template:rnn_cell) 'decoder_1' out_type=Data(shape=(640,), time_dim_axis=None, batch_shape_meta=[B,F|640]) (construction stack 'decoder_2')>,
 <_TemplateLayer(GenericAttentionLayer)(:template:generic_attention) 'att0' out_type=Data(shape=(384,), time_dim_axis=None, batch_shape_meta=[B,F|384]) (construction stack 'decoder_1')>,
 <_TemplateLayer(SoftmaxOverSpatialLayer)(:template:softmax_over_spatial) 'att_weights' out_type=Data(shape=(1, None), time_dim_axis=2, feature_dim_axis=1, available_for_inference=False, batch_shape_meta=[B,F|1,T|'time:var:extern_data:classes']) (construction stack 'att0')>,
 <_TemplateLayer(LinearLayer)(:template:linear) 'att_energy' out_type=Data(shape=(None, 1), batch_dim_axis=1, available_for_inference=False, batch_shape_meta=[T|'time:var:extern_data:classes',B,F|1]) (construction stack 'att_weights')>,
 <_TemplateLayer(ActivationLayer)(:template:activation) 'att_energy_tanh' out_type=Data(shape=(None, 128), batch_dim_axis=1, available_for_inference=False, batch_shape_meta=[T|'time:var:extern_data:classes',B,F|128]) (construction stack 'att_energy')>,
 <_TemplateLayer(CombineLayer)(:template:combine) 'att_energy_in' out_type=Data(shape=(None, 128), batch_dim_axis=1, available_for_inference=False, batch_shape_meta=[T|'time:var:extern_data:classes',B,F|128]) (construction stack 'att_energy_tanh')>,
 <_TemplateLayer(LinearLayer)(:template:linear) 's_transformed' out_type=Data(shape=(128,), time_dim_axis=None, batch_shape_meta=[B,F|128]) (construction stack 'att_energy_in')>,
 <_TemplateLayer(LinearLayer)(:template:linear) 'location_feedback_transformed' out_type=Data(shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|?,F|128]) (construction stack 'att_energy_in')>,
 <_TemplateLayer(ConvLayer)(:template:conv) 'convolved_att' out_type=Data(shape=(None, 32), available_for_inference=False, batch_shape_meta=[B,T|?,F|32]) (construction stack 'location_feedback_transformed')>,
 <_TemplateLayer(PadLayer)(:template:pad) 'feedback_pad_right' out_type=Data(shape=(None, 1), time_dim_axis=None, available_for_inference=False, batch_shape_meta=[B,'time:var:extern_data:classes',F|1]) (construction stack 'convolved_att')>,
 <_TemplateLayer(PadLayer)(:template:pad) 'feedback_pad_left' out_type=Data(shape=(None, 1), time_dim_axis=None, available_for_inference=False, batch_shape_meta=[B,'time:var:extern_data:classes',F|1]) (construction stack 'feedback_pad_right')>,
 <_TemplateLayer(EvalLayer)(:template:eval) 'accum_att_weights' out_type=Data(shape=(None, 1), time_dim_axis=None, available_for_inference=False, batch_shape_meta=[B,'time:var:extern_data:classes',F|1]) (construction stack 'feedback_pad_left')>,
 <_TemplateLayer(EvalLayer)(:template:eval) 'entropy' out_type=Data(shape=(1, None), time_dim_axis=2, feature_dim_axis=1, available_for_inference=False, batch_shape_meta=[B,F|1,T|'time:var:extern_data:classes']) (construction stack None)>]
And these finished layers:
[<_TemplateLayer(ChoiceLayer)(:template:choice) 'choice' out_type=Data(shape=(240,), time_dim_axis=None, batch_shape_meta=[B,F|240]) (construction stack 'pre_slice')>,
 <_TemplateLayer(LinearLayer)(:template:linear) 'pre_net_layer_1' out_type=Data(shape=(128,), time_dim_axis=None, batch_shape_meta=[B,F|128]) (construction stack 'pre_net_layer_2')>,
 <_TemplateLayer(LinearLayer)(:template:linear) 'pre_net_layer_2' out_type=Data(shape=(64,), time_dim_axis=None, batch_shape_meta=[B,F|64]) (construction stack 'pre_net_layer_2_out')>,
 <_TemplateLayer(DropoutLayer)(:template:dropout) 'pre_net_layer_2_out' out_type=Data(shape=(64,), time_dim_axis=None, batch_shape_meta=[B,F|64]) (construction stack 'decoder_1')>,
 <_TemplateLayer(SliceLayer)(:template:slice) 'pre_slice' out_type=Data(shape=(80,), time_dim_axis=None, batch_shape_meta=[B,F|80]) (construction stack 'pre_net_layer_1')>]
And these recent changes in the last loop iteration:
[(Data(name='location_feedback_transformed_output', shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|'time:var:extern_data:classes',F|128]),
  Data(name='location_feedback_transformed_output', shape=(None, 128), available_for_inference=False, batch_shape_meta=[B,T|?,F|128]))]

```

**Comment by albertz** [comment]
What is the layer dict of `location_feedback_transformed`?

**Comment by albertz** [comment]
Also for `convolved_att`.

**Comment by JackTemaki** [comment]
```
                         'feedback_pad_left': { 'axes': 's:0',
                                                'class': 'pad',
                                                'from': ['prev:accum_att_weights'],
                                                'mode': 'constant',
                                                'padding': ((15, 0),),
                                                'value': 1},
                         'feedback_pad_right': { 'axes': 's:0',
                                                 'class': 'pad',
                                                 'from': ['feedback_pad_left'],
                                                 'mode': 'constant',
                                                 'padding': ((0, 15),),
                                                 'value': 0},
                         'convolved_att': { 'L2': 0.01,
                                            'activation': None,
                                            'class': 'conv',
                                            'filter_size': (31,),
                                            'from': ['feedback_pad_right'],
                                            'n_out': 32,
                                            'padding': 'valid'},
                         'location_feedback_transformed': { 'L2': 0.01,
                                                            'activation': None,
                                                            'class': 'linear',
                                                            'dropout': 0.1,
                                                            'from': ['convolved_att'],
                                                            'n_out': 128,
                                                            'with_bias': False},

**Comment by JackTemaki** [comment]
When the convolution layer is removed, the code works fine, so the error might be somewhere there. I will investigate further.

**Comment by JackTemaki** [comment]
The issue seems to be that the ConvLayer can not automatically keep the time-axis size of the previous layer. One possible reason is the lacking time axis definition in `def _get_out_type_from_opts`, so that during the check at
```
matches_var_dim_pattern, TFUtil.py:1587
_post_init_output, TFNetworkLayer.py:356
_base_get_out_data_from_opts, TFNetworkLayer.py:326
get_out_data_from_opts, TFNetworkLayer.py:226
```
which is
```
 self.time_dim_axis_excluding_batch != other.time_dim_axis_excluding_batch:
      return False
```

the time_axis does not match, and the size is not set properly here:
```
      elif common_source and common_source.matches_var_dim_pattern(output):
        output.size_placeholder = common_source.size_placeholder.copy()
```

When setting the `time_dim_axis` explicitely in ConvLayer._get_out_type_from_opts
```
    return {
      "dim": n_out,
      "shape": shape,
      "batch_dim_axis": 0,
      "feature_dim_axis": feature_dim_axis,
>>>   "time_dim_axis": data.time_dim_axis,
      "sparse": False}
```
The error disappears. The quesiton is if this should be done this way, and if this can lead to conflicts.

**Comment by albertz** [comment]
What is `self.time_dim_axis_excluding_batch`, and what is `other.time_dim_axis_excluding_batch`?
Do you have the test case ready so I can easily try out?

**Comment by albertz** [comment]
Why is `other.time_dim_axis_excluding_batch` None? Normally if you do not set it explicitly, it will try to use a reasonable default, and e.g. if you have `shape=(None, ...)`, it would (should; I think...) automatically set it.

**Comment by JackTemaki** [comment]
The Issue was fixed with commit 4c696132ab86c2864cbd420ec63c778618472009
